rules/DeadCode/Rector/StmtsAwareInterface/RemoveJustPropertyFetchRector.php

```diff
@@ -90,158 +74,50 @@ public function refactor(Node $node): ?Node
-            $node,
-            $currentStmtKey,
-            $variableUsages,
-            $variableToPropertyAssign->getPropertyFetch()
-        );
-    }
-
-    /**
-     * @param Stmt[] $followingStmts
-     */
-    private function isFollowingStatementStaticClosure(array $followingStmts): bool
-    {
-        if ($followingStmts === []) {
-            return false;
-        }
-
-        return $followingStmts[0] instanceof Expression
-            && $followingStmts[0]->expr instanceof Closure
-            && $followingStmts[0]->expr->static;
-    }
-
-    /**
-     * @param Variable[] $variableUsages
-     */
-    private function replaceVariablesWithPropertyFetch(
-        StmtsAwareInterface $stmtsAware,
-        int $currentStmtsKey,
-        array $variableUsages,
-        PropertyFetch $propertyFetch
-    ): StmtsAwareInterface {
-        // remove assign node
-        unset($stmtsAware->stmts[$currentStmtsKey]);
-
-        $this->traverseNodesWithCallable(
-            $stmtsAware,
-            function (Node $node) use ($variableUsages, $propertyFetch): ?PropertyFetch {
-                if (! in_array($node, $variableUsages, true)) {
-                    return null;
-                }
-
-                $parentNode = $node->getAttribute(AttributeKey::PARENT_NODE);
-                if ($parentNode instanceof ClosureUse) {
-                    // remove closure use which will be replaced by a property fetch
-                    $this->removeNode($parentNode);
-
-                    return null;
-                }
-
-                return $propertyFetch;
-            }
-        );
-
-        return $stmtsAware;
-    }
-
-    /**
-     * @param Variable[] $variableUsages
-     * @return Variable[]
-     */
-    private function filterOutReferencedVariableUsages(array $variableUsages): array
-    {
-        return array_filter($variableUsages, function (Variable $variable): bool {
-            $variableUsageParent = $variable->getAttribute(AttributeKey::PARENT_NODE);
+            $assignPhpDocInfo = $this->phpDocInfoFactory->createFromNodeOrEmpty($previousStmt);
 
-            if ($variableUsageParent instanceof Arg) {
-                $variableUsageParent = $variableUsageParent->getAttribute(AttributeKey::PARENT_NODE);
-            }
-
-            // skip nested property fetch, the assign is for purpose of named variable
-            if ($variableUsageParent instanceof PropertyFetch) {
-                return false;
+            // there is a @var tag on purpose, keep the assign
+            if ($assignPhpDocInfo->getVarTagValueNode() instanceof VarTagValueNode) {
+                continue;
             }
 
-            // skip, as assign can be used in a loop
-            $parentWhile = $this->betterNodeFinder->findParentType($variable, While_::class);
-            if ($parentWhile instanceof While_) {
-                return false;
-            }
+            unset($node->stmts[$key - 1]);
```

Comment:
@samsonasik I'd suspect this line. The node tree needs keys to be reset an in order, e.g. 1, 2, 3... Here one is removed, so order is 1, 3.

This could be place to reset keys or debug.

Reply:
I found that reindex stmts stmt_key is needed, see PR:

- https://github.com/rectorphp/rector-src/pull/3941

that seems resolve it.